jigglypuffRL/classicalrl/vpg/vpg.py

```diff
@@ -0,0 +1,217 @@
+        self.clip_param = clip_param
+        self.actor_batch_size = actor_batch_size
+        self.epochs = epochs
+        self.lr_policy = lr_policy
+        self.lr_value = lr_value
+        self.tensorboard_log = tensorboard_log
+        self.seed = seed
+        self.render = render
+        self.policy_copy_interval = policy_copy_interval
+
+        # Assign device
+        if "cuda" in device and torch.cuda.is_available():
+            self.device = torch.device(device)
+        else:
+            self.device = torch.device("cpu")
+
+        # Assign seed
+        if seed is not None:
+            torch.manual_seed(seed)
+            torch.backends.cudnn.deterministic = True
+            torch.backends.cudnn.benchmark = False
+            np.random.seed(seed)
+            self.env.seed(seed)
+
+        # init writer if tensorboard
+        self.writer = None
+        if self.tensorboard_log is not None:
+            from torch.utils.tensorboard import SummaryWriter
+
+            self.writer = SummaryWriter(log_dir=self.tensorboard_log)
+
+        self.create_model()
+
+    def create_model(self):
+        # Instantiate networks and optimizers
+        self.policy_fn = get_policy_from_name(self.policy)(self.env).to(self.device)
+        self.value_fn = get_value_from_name(self.value)(self.env).to(self.device)
+        
+        self.optimizer_policy = opt.Adam(self.policy_fn.parameters(), lr=self.lr_policy)
+        self.optimizer_value = opt.Adam(self.value_fn.parameters(), lr=self.lr_value)
+
+    def select_action(self, s):
+        state = torch.from_numpy(s).float().to(self.device)
+
+        # create distribution based on policy_fn output
+        action, c = self.policy_fn.sample_action(Variable(state))
+        val = self.value_fn(Variable(state))
+
+        # store policy probs and value function for current traj
+        self.policy_fn.policy_hist = torch.cat(
+            [
+                self.policy_fn.policy_hist,
+                c.log_prob(action).unsqueeze(0),
+            ]
+        )
+
+        self.value_fn.value_hist = torch.cat([self.value_fn.value_hist, val])
+
+        return action
+
+    # get clipped loss for single trajectory (episode)
+    def get_traj_loss(self):
+        R = 0
+        returns = []
+
+        # calculate discounted return
+        for r in self.policy_fn.traj_reward[::-1]:
+            R = r + self.gamma * R
+            returns.insert(0, R)
+
+        # advantage estimation
+        returns = torch.FloatTensor(returns).to(self.device)
+        A = Variable(returns) - Variable(self.value_fn.value_hist)
+
+        # compute policy and value loss
+        clipping = (
+            torch.clamp(self.policy_fn.policy_hist, 1 - self.clip_param, 1 + self.clip_param)
+            .mul(A)
+            .to(self.device)
+        )
+
+        loss_policy = (
+            torch.mean(torch.min(torch.mul(self.policy_fn.policy_hist, A), clipping)).mul(-1).unsqueeze(0)
+        )
```

Comment:
The policy loss calculation appears to be incorrect. In standard VPG, there is no clipping of policies. Rather, the loss should simply be:
```python
loss_policy = torch.sum(
  torch.mul(self.policy_fn.policy_hist, A)
).mul(-1).unsqueeze(0)
```
Note that I've replaced mean with sum since VPG averages loss only with respect to the number of trajectories. (Refer https://spinningup.openai.com/en/latest/algorithms/vpg.html)